cmd/update/matchable_attribute_util.go

```diff
@@ -3,45 +3,196 @@ package update
-	updater ext.AdminUpdaterExtInterface, mcDecorator sconfig.MatchableAttributeDecorator, dryRun bool) error {
-	matchingAttr := mcDecorator.Decorate()
-	if len(workflowName) > 0 {
-		// Update the workflow attribute using the admin.
-		if dryRun {
-			fmt.Printf("skipping UpdateWorkflowAttributes request (dryRun)\n")
-		} else {
-			err := updater.UpdateWorkflowAttributes(ctx, project, domain, workflowName, matchingAttr)
-			if err != nil {
-				return err
-			}
-		}
-		fmt.Printf("Updated attributes from %v project and domain %v and workflow %v\n", project, domain, workflowName)
-	} else {
-		// Update the project domain attribute using the admin.
-		if dryRun {
-			fmt.Printf("skipping UpdateProjectDomainAttributes request (dryRun)\n")
-		} else {
-			if len(domain) == 0 {
-				err := updater.UpdateProjectAttributes(ctx, project, matchingAttr)
-				if err != nil {
-					return err
-				}
-				fmt.Printf("Updated attributes from %v project\n", project)
-			} else {
-				err := updater.UpdateProjectDomainAttributes(ctx, project, domain, matchingAttr)
-				if err != nil {
-					return err
-				}
-				fmt.Printf("Updated attributes from %v project and domain %v\n", project, domain)
-			}
-		}
+func DecorateAndUpdateMatchableAttr(
+	ctx context.Context,
+	cmdCtx cmdCore.CommandContext,
+	project, domain, workflow string,
+	resourceType admin.MatchableResource,
+	attributeDecorator sconfig.MatchableAttributeDecorator,
+	dryRun bool,
+	force bool,
+) error {
+	if project == "" {
+		return fmt.Errorf("project is required")
+	}
+	if domain == "" && workflow != "" {
+		return fmt.Errorf("domain is required")
+	}
+
+	switch {
+	case workflow != "":
+		return updateWorkflowMatchableAttributes(ctx, cmdCtx, project, domain, workflow, resourceType, attributeDecorator, dryRun, force)
+	case domain != "":
+		return updateProjectDomainMatchableAttributes(ctx, cmdCtx, project, domain, resourceType, attributeDecorator, dryRun, force)
+	default:
+		return updateProjectMatchableAttributes(ctx, cmdCtx, project, resourceType, attributeDecorator, dryRun, force)
+	}
+}
+
+func updateProjectMatchableAttributes(
+	ctx context.Context,
+	cmdCtx cmdCore.CommandContext,
+	project string,
+	resourceType admin.MatchableResource,
+	attributeDecorator sconfig.MatchableAttributeDecorator,
+	dryRun bool,
+	force bool,
+) error {
+	if project == "" {
+		panic("project is empty")
+	}
+
+	response, err := cmdCtx.AdminFetcherExt().FetchProjectAttributes(ctx, project, resourceType)
+	if err != nil && !ext.IsNotFoundError(err) {
+		return fmt.Errorf("update project %s matchable attributes: could not fetch attributes: %w", project, err)
+	}
+
+	oldMatchingAttributes := response.GetAttributes().GetMatchingAttributes()
+	newMatchingAttributes := attributeDecorator.Decorate()
+
+	patch, err := DiffAsYaml("before", "after", oldMatchingAttributes.GetTarget(), newMatchingAttributes.GetTarget())
+	if err != nil {
+		panic(err)
```

Comment:
I made a similar comment elsewhere. I didn't realize/confirm that the inputs here were fully validated and any failures were purely internal. If so, agreed.

Reply:
Panic prints a stack trace which may not be something we want to show in user facing interface even in case of bugs and hence the concern to have panic.

The main routine, handles any errors and returns with exit 1. 

If we print the error is it not sufficient for us to debug if the user reports the error? Do we need the stack trace to be shown the user and shared with us to debug any issues?